src/ngraph/runtime/gpu/cublas_emitter.cpp

```diff
@@ -0,0 +1,237 @@
+
+runtime::gpu::CUBLASEmitter::CUBLASEmitter(GPUPrimitiveEmitter* emitter, GPURuntimeContext* ctx)
+    : m_primitive_emitter(emitter)
+{
+    m_ctx = ctx;
+}
+
+size_t runtime::gpu::CUBLASEmitter::build_dot(const element::Type& dtype,
+                                              const Shape& arg0_shape,
+                                              const Shape& arg1_shape,
+                                              const Shape& out_shape,
+                                              size_t reduction_axes)
+{
+    std::stringstream ss;
+    ss << "dot_op"
+       << "_dtype_" << dtype.c_type_string() << "_reduction_axes_count_" << reduction_axes;
+    std::string hash = ss.str() + "_i_" + join(arg0_shape, "_") + "_i_" + join(arg1_shape, "_") +
+                       "_o_" + join(out_shape, "_");
+
+    size_t primitive_index = m_primitive_emitter->lookup(hash);
+    if (primitive_index != std::numeric_limits<size_t>::max())
+    {
+        return primitive_index;
+    }
+
+    std::unique_ptr<gpu::primitive> dot;
+    if (arg0_shape.empty() || arg1_shape.empty())
+    {
+        auto& second = (arg0_shape.empty() ? arg1_shape : arg0_shape);
+        size_t count = shape_size(second);
+
+        size_t firstIndex = (arg0_shape.empty() ? 0 : 1);
+        size_t secondIndex = (arg0_shape.empty() ? 1 : 0);
+
+        dot.reset(new gpu::primitive{[=](void** inputs, void** outputs) {
+            CUBLAS_SAFE_CALL(cublasScopy(*m_ctx->cublas_handle,
+                                         count,
+                                         static_cast<const float*>(inputs[secondIndex]),
+                                         1,
+                                         static_cast<float*>(outputs[0]),
+                                         1));
+            CUBLAS_SAFE_CALL(cublasSscal(*m_ctx->cublas_handle,
+                                         count,
+                                         static_cast<const float*>(inputs[firstIndex]),
+                                         static_cast<float*>(outputs[0]),
+                                         1));
+            debug_sync();
+        }});
+
+        return this->m_primitive_emitter->get_primitive_index(dot, hash);
+    }
+
+    if (shape_size(arg0_shape) == 0 || shape_size(arg1_shape) == 0)
+    {
+        size_t elemSize = shape_size(out_shape) * dtype.size();
+        dot.reset(new gpu::primitive{[=](void** inputs, void** outputs) {
+            runtime::gpu::cuda_memset(outputs[0], 0, elemSize);
+            debug_sync();
+        }});
+
+        return this->m_primitive_emitter->get_primitive_index(dot, hash);
+    }
+    // case that can be treat as dot1d
+    if ((arg0_shape.size() == arg1_shape.size()) && (arg0_shape.size() == reduction_axes))
+    {
+        for (int i = 0; i < arg0_shape.size(); i++)
+        {
+            if (arg0_shape[i] != arg1_shape[i])
+            {
+                throw std::invalid_argument("arg0 and arg1 shape does not match for dot.");
```

Comment:
You should probably add more information to this error, like what are the shapes? Would be nice to know what the node is, but that requires the function signature to change. If you get this error you want to know where and why.

Reply:
I resolved this. The error msg now looks something like this:
"arg0, arg1 with Shape{<>,<>}, Shape{<>, <>} respectively, at Node Dot_<>, do not match for dot op"